Add utils functions

mary-cleaton opened the pull request:
## Description
Add utility functions and their unit tests from old repository.

## Type of change

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

Commit:
feat: add utils functions and their tests

Commit:
docs: update spark version number in readme

Commit:
docs: update changelog

grthomson commented on tests/utils/test_functions.py:
```diff
@@ -0,0 +1,302 @@
+
+
+#  # Arrange
+#  test_input = spark.createDataFrame(
```
Currently marked to be skipped, but this commented out section made me wonder if you were coming back to this?

L13 also "function updated, test not"

mary-cleaton replied:
I plan to come back to it, but think we can release the code before revisiting it. I hadn't updated the function now - the update was done a while ago back when the code was in development. It just wasn't spotted as the function name was changed so the test didn't fail.

mary-cleaton replied:
Added this as an issue - #22.

grthomson commented on tests/utils/test_functions.py:
```diff
@@ -0,0 +1,302 @@
+    pass
+
+
+#  # Arrange
```
Should this commented out code be kept for now?

mary-cleaton replied:
Yeah, my feeling with both of these was to come back to them in a future release as they don't impede the code from working, they just reduce the test coverage.

mary-cleaton replied:
Added this as an issue - #21.

Commit:
style: format module docstring

Commit:
Merge branch 'develop' into feat/utils

Commit:
ci: run pre-commits

grthomson commented:
Unit tests all pass with no errors with pytest in terminal. 

But PR Workflow for unit tests currently failing. Looks like the root cause is pytest looking for unit tests in the wrong place?

https://github.com/ONSdigital/scalelink/actions/runs/16990917250/job/48169856107?pr=18

Commit:
fix: change function and test script names
Pytest refused to run when multiple function scripts were named  and multiple test scripts were named .

mary-cleaton commented:
> Unit tests all pass with no errors with pytest in terminal.
> 
> But PR Workflow for unit tests currently failing. Looks like the root cause is pytest looking for unit tests in the wrong place?
> 
> https://github.com/ONSdigital/scalelink/actions/runs/16990917250/job/48169856107?pr=18

Found the root cause: pytest didn't like that all the function scripts had the same name and all the test scripts had the same name. I've now given them different names and it's fixed the tests so they run.

grthomson reviewed (approved):
Unit test Actions workflow now passing, as well as pytest in terminal. So approving this. :)
